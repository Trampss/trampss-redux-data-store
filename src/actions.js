const scope = action => `@trampss/${action.toUpperCase()}`

export const SET = prefix => scope(`SET_${prefix}`)
export const set = prefix => payload => ({ type: SET(prefix), payload })

export const RESET = prefix => scope(`RESET_${prefix}`)
export const reset = prefix => () => ({ type: RESET(prefix) })

export const ADD = prefix => scope(`ADD_${prefix}`)
export const add = prefix => payload => ({ type: ADD(prefix), payload })

export const UPDATE = prefix => scope(`UPDATE_${prefix}`)
export const update = prefix => payload => ({ type: UPDATE(prefix), payload })

export const REMOVE = prefix => scope(`REMOVE_${prefix}`)
export const remove = prefix => key => ({ type: REMOVE(prefix), payload: key })
