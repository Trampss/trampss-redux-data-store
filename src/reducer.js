import { keyBy, without, uniq, omit } from 'lodash'
import { SET, ADD, DEL, RESET } from './actions'

export const initState = { data: {}, keys: [], array: [], nb: 0, initialized: false }

export default key => prefix =>
  (state = initState, { type = 'UNKONWN', payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: keyBy(payload, key),
          keys: payload.map(element => element[key]),
          array: payload,
          nb: payload.length,
          initialized: true,
        }
      case ADD(prefix):
        return {
          ...state,
          data: { ...state.data, [payload[key]]: payload },
          keys: uniq([...state.keys, payload[key]]),
          array: [...state.array, payload],
          nb: state.keys.length + 1,
          initialized: true,
        }
      case DEL(prefix):
        return {
          ...state,
          data: omit(state.data, [payload]),
          keys: without(state.keys, payload),
          array: state.array ? state.array.filter(o => o[key] !== payload) : [],
          nb: state.keys.length - 1,
        }
      case RESET(prefix):
        return initState
      default:
        return state
    }
  }
