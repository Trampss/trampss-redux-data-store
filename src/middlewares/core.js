import { keyBy, without, uniq, omit } from 'lodash'
import { SET, ADD, REMOVE, RESET } from '../actions'
import { initState } from '../reducer'

const coreReducer = key => prefix =>
  (state = initState, { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: keyBy(payload, key),
          keys: uniq(payload.map(element => element[key])),
          array: payload,
          initialized: true,
        }
      case ADD(prefix): {
        const sameKey = state.array.find(o => o[key] === payload[key])
        let array
        if (sameKey === undefined) {
          array = [...state.array, payload]
        } else {
          array = state.array.map(
            o => (o[key] === payload[key] ? payload : o),
          )
        }

        return {
          ...state,
          data: { ...state.data, [payload[key]]: payload },
          keys: uniq([...state.keys, payload[key]]),
          array,
          initialized: true,
        }
      }
      case REMOVE(prefix):
        return {
          ...state,
          data: omit(state.data, [payload]),
          keys: without(state.keys, payload),
          array: state.array ? state.array.filter(o => o[key] !== payload) : [],
        }
      case RESET(prefix):
        return initState
      default:
        return state
    }
  }

export default key => prefix => (ctx = {}) => ({
  ...ctx,
  state: coreReducer(key)(prefix)(ctx.state, ctx.action),
})
