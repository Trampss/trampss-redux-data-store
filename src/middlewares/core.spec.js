/* eslint-env jest */
import { set, add, update, remove, reset } from '../actions'
import { initState } from '../reducer'
import core from './core'

const prefix = 'testPrefix'
const Element = code => ({ code, some: 'other', infos: code })
const state = {
  data: {
    elm2: Element('elm2'),
    elm1: Element('elm1'),
    elm3: Element('elm3'),
  },
  keys: ['elm2', 'elm1', 'elm3'],
  array: [Element('elm2'), Element('elm1'), Element('elm3')],
  initialized: false,
}

describe('middlewares/core', () => {
  // Use the factory to create a new reducer named 'testPrefix'
  // It uses 'code' as key in givent elements
  const testPrefix = core('code')(prefix)

  it('should initialize', () => expect(testPrefix()).toMatchSnapshot())

  it('should add element [elm4] on initial store', () => expect(
    testPrefix({
      state: initState,
      action: add(prefix)(Element('elm4')),
    }),
  ).toMatchSnapshot())

  it('should add element [elm4]', () => expect(
    testPrefix({
      state,
      action: add(prefix)(Element('elm4')),
    }),
  ).toMatchSnapshot())

  it('should set elements [elm1,elm2]', () => expect(
    testPrefix({
      state,
      action: set(prefix)([Element('elm1'), Element('elm2')]),
    }),
  ).toMatchSnapshot())

  it('should remove element [elm1]', () => expect(
    testPrefix({
      state,
      action: remove(prefix)('elm1'),
    }),
  ).toMatchSnapshot())

  it('should reset state', () => expect(
    testPrefix({
      state,
      action: reset(prefix)(),
    }),
  ).toMatchSnapshot())

  it('should not duplicate object with same key [elm2]', () => expect(
    testPrefix({
      state,
      action: add(prefix)(Element('elm2')),
    }),
  ).toMatchSnapshot())

  it('should update an element [elm3]', () => expect(
    testPrefix({
      state,
      action: update(prefix)({ ...Element('elm3'), some: 'other UPDATE' }),
    }),
  ).toMatchSnapshot())

  it('should update an element not present [elm12]', () => expect(
    testPrefix({
      state,
      action: update(prefix)(Element('elm12')),
    }),
  ).toMatchSnapshot())
})
