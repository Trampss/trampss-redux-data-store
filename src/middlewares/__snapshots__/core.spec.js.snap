// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/core should add element [elm4] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
    },
    "type": "@trampss/ADD_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
      "elm4": Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
      "elm4",
    ],
  },
}
`;

exports[`middlewares/core should add element [elm4] on initial store 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm4",
      "infos": "elm4",
      "some": "other",
    },
    "type": "@trampss/ADD_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    ],
    "data": Object {
      "elm4": Object {
        "code": "elm4",
        "infos": "elm4",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm4",
    ],
  },
}
`;

exports[`middlewares/core should initialize 1`] = `
Object {
  "state": Object {
    "array": Array [],
    "data": Object {},
    "initialized": false,
    "keys": Array [],
  },
}
`;

exports[`middlewares/core should not duplicate object with same key [elm2] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "type": "@trampss/ADD_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/core should remove element [elm1] 1`] = `
Object {
  "action": Object {
    "payload": "elm1",
    "type": "@trampss/REMOVE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/core should reset state 1`] = `
Object {
  "action": Object {
    "type": "@trampss/RESET_TESTPREFIX",
  },
  "state": Object {
    "array": Array [],
    "data": Object {},
    "initialized": false,
    "keys": Array [],
  },
}
`;

exports[`middlewares/core should set elements [elm1,elm2] 1`] = `
Object {
  "action": Object {
    "payload": Array [
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
    ],
    "type": "@trampss/SET_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
    },
    "initialized": true,
    "keys": Array [
      "elm1",
      "elm2",
    ],
  },
}
`;

exports[`middlewares/core should update an element [elm3] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm3",
      "infos": "elm3",
      "some": "other UPDATE",
    },
    "type": "@trampss/UPDATE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other UPDATE",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other UPDATE",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;

exports[`middlewares/core should update an element not present [elm12] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm12",
      "infos": "elm12",
      "some": "other",
    },
    "type": "@trampss/UPDATE_TESTPREFIX",
  },
  "state": Object {
    "array": Array [
      Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    ],
    "data": Object {
      "elm1": Object {
        "code": "elm1",
        "infos": "elm1",
        "some": "other",
      },
      "elm2": Object {
        "code": "elm2",
        "infos": "elm2",
        "some": "other",
      },
      "elm3": Object {
        "code": "elm3",
        "infos": "elm3",
        "some": "other",
      },
    },
    "initialized": false,
    "keys": Array [
      "elm2",
      "elm1",
      "elm3",
    ],
  },
}
`;
